src/main/java/org/venice/beachfront/bfapi/services/OAuthService.java

```diff
@@ -0,0 +1,110 @@
+import java.util.UUID;
+
+import org.joda.time.DateTime;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClientResponseException;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+import org.venice.beachfront.bfapi.model.UserProfile;
+import org.venice.beachfront.bfapi.model.exception.UserException;
+import org.venice.beachfront.bfapi.model.oauth.AccessTokenRequestBody;
+import org.venice.beachfront.bfapi.model.oauth.AccessTokenResponseBody;
+import org.venice.beachfront.bfapi.model.oauth.ProfileResponseBody;
+
+@Service
+public class OAuthService {
+	@Value("${DOMAIN}")
+	private String domain;
+
+	@Value("${oauth.token-url}")
+	private String oauthTokenUrl;
+
+	@Value("${oauth.profile-url")
+	private String oauthProfileUrl;
+
+	@Value("${OAUTH_CLIENT_ID}")
+	private String oauthClientId;
+
+	@Value("${OAUTH_SECRET}")
+	private String oauthClientSecret;
+
+	@Autowired
+	private RestTemplate restTemplate;
+
+	@Autowired
+	private UserProfileService userProfileService;
+
+	public String getOauthRedirectUri() {
+		return UriComponentsBuilder.newInstance().host("bf-api." + this.domain).pathSegment("oauth", "callback").build().toUri().toString();
+	}
+
+	public String requestAccessToken(String authCode) throws UserException {
+		AccessTokenRequestBody body = new AccessTokenRequestBody("authorization_code", authCode, this.getOauthRedirectUri());
+		HttpHeaders headers = new HttpHeaders();
+		headers.set("Authorization", this.createTokenAuthHeader());
+		HttpEntity<AccessTokenRequestBody> entity = new HttpEntity<>(body, headers);
+
+		ResponseEntity<AccessTokenResponseBody> response;
+		try {
+			response = this.restTemplate.exchange(this.oauthTokenUrl, HttpMethod.POST, entity, AccessTokenResponseBody.class);
+		} catch (RestClientResponseException ex) {
+			throw new UserException("Failed requesting OAuth access token", ex, HttpStatus.UNAUTHORIZED);
+		}
+
+		return response.getBody().getAccessToken();
+	}
+
+	public ProfileResponseBody requestOAuthProfile(String accessToken) throws UserException {
+		HttpHeaders headers = new HttpHeaders();
+		headers.set("Authorization", "Bearer " + accessToken);
+		HttpEntity<Object> entity = new HttpEntity<>(null, headers);
+
+		ResponseEntity<ProfileResponseBody> response;
+		try {
+			response = this.restTemplate.exchange(this.oauthProfileUrl, HttpMethod.GET, entity, ProfileResponseBody.class);
+		} catch (RestClientResponseException ex) {
+			throw new UserException("Failed requesting OAuth access token", ex, HttpStatus.UNAUTHORIZED);
```

Comment:
Filip: Is this always going to be a 401?  There could be an incorrect URL upstream or something, in which case I don't think a 401 applies. Do we want to look at the status code returned by the Exchange method and use that?

Reply:
We probably want a little more granular error handling. If the status is 4xx, proxy that. If the status is 5xx, return 502 Bad Gateway with some extra details and logging. If anything else, 500.